tutorials/src/bin/basic-tutorial-8.rs

```diff
@@ -0,0 +1,260 @@
+use gst_audio::AudioInfo;
+extern crate gstreamer_app as gst_app;
+use gst_app::{AppSrc, AppSink};
+extern crate glib;
+use glib::source::{SourceId, idle_add, source_remove};
+
+const CHUNK_SIZE: usize = 1024; // Amount of bytes we are sending in each buffer
+const SAMPLE_RATE: u32 = 44100; // Samples per second we are sending
+
+#[derive(Debug)]
+struct CustomData {
+    source_id: Option<SourceId>,
+
+    num_samples: u64, // Number of samples generated so far (for timestamp generation)
+    // For waveform generation
+    a: f64,
+    b: f64,
+    c: f64,
+    d: f64,
+
+    appsrc: AppSrc,
+    appsink: AppSink,
+}
+
+impl CustomData {
+    fn new(appsrc: AppSrc, appsink: AppSink) -> CustomData {
+        CustomData {
+            source_id: None,
+            num_samples: 0,
+            a: 0.0,
+            b: 1.0,
+            c: 0.0,
+            d: 1.0,
+            appsrc: appsrc,
+            appsink: appsink,
+        }
+    }
+}
+
+fn main() {
+    // Initialize GStreamer
+    if let Err(err) = gst::init() {
+        eprintln!("Failed to initialize Gst: {}", err);
+        return;
+    }
+
+    let app_source = gst::ElementFactory::make("appsrc", "audio_source").unwrap();
+    let tee = gst::ElementFactory::make("tee", "tee").unwrap();
+    let audio_queue = gst::ElementFactory::make("queue", "audio_queue").unwrap();
+    let audio_convert1 = gst::ElementFactory::make("audioconvert", "audio_convert1").unwrap();
+    let audio_resample = gst::ElementFactory::make("audioresample", "audio_resample").unwrap();
+    let audio_sink = gst::ElementFactory::make("autoaudiosink", "audio_sink").unwrap();
+    let video_queue = gst::ElementFactory::make("queue", "video_queue").unwrap();
+    let audio_convert2 = gst::ElementFactory::make("audioconvert", "audio_convert2").unwrap();
+    let visual = gst::ElementFactory::make("wavescope", "visual").unwrap();
+    let video_convert = gst::ElementFactory::make("videoconvert", "video_convert").unwrap();
+    let video_sink = gst::ElementFactory::make("autovideosink", "video_sink").unwrap();
+    let app_queue = gst::ElementFactory::make("queue", "app_queue").unwrap();
+    let app_sink = gst::ElementFactory::make("appsink", "app_sink").unwrap();
+
+    let pipeline = gst::Pipeline::new("test-pipeline");
+
+    visual.set_property_from_str("shader", "none");
+    visual.set_property_from_str("style", "lines");
+
+    // configure appsrc
+    let info = AudioInfo::new(gst_audio::AudioFormat::S16le, SAMPLE_RATE, 1)
+        .build()
+        .unwrap();
+    let audio_caps = info.to_caps().unwrap();
+    app_source.set_property("caps", &audio_caps).unwrap();
+    app_source.set_property_from_str("format", "time");
+
+    let appsrc = app_source.clone().dynamic_cast::<AppSrc>().expect(
+        "Source element is expected to be an appsrc!",
+    );
+
+    let appsink = app_sink.clone().dynamic_cast::<AppSink>().expect(
+        "Sink element is expected to be an appsink!",
+    );
```

Comment:
Maybe instead of having two variables that have almost the same name (appsrc and app_source), just cast here and use that in the future?

Reply:
Done. I just had to move code around a bit to do so.